crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::exit_codes::{CInputError, ExitCode};
+use num_complex::Complex64;
+use qiskit_accelerate::sparse_observable::{BitTerm, SparseObservable, SparseTermView};
+
+/// @ingroup QkSparseTerm
+/// A term in a [SparseObservable].
+///
+/// This term does not own the data, but only contains pointers (and a size) to the
+/// bit terms, indices and the coefficient.
+#[repr(C)]
+pub struct CSparseTerm {
+    coeff: *mut Complex64,
+    len: usize,
+    bit_terms: *mut BitTerm,
+    indices: *mut u32,
+    num_qubits: u32,
+}
+
+impl TryFrom<&CSparseTerm> for SparseTermView<'_> {
+    type Error = CInputError;
+
+    fn try_from(value: &CSparseTerm) -> Result<Self, Self::Error> {
+        if value.bit_terms.is_null() || value.indices.is_null() {
+            return Err(CInputError::NullPointerError);
+        }
+
+        if !value.bit_terms.is_aligned() || !value.indices.is_aligned() {
+            return Err(CInputError::AlignmentError);
+        }
+
+        let bit_terms = unsafe { ::std::slice::from_raw_parts(value.bit_terms, value.len) };
+        let indices = unsafe { ::std::slice::from_raw_parts(value.indices, value.len) };
+
+        Ok(SparseTermView {
+            num_qubits: value.num_qubits,
+            coeff: unsafe { *value.coeff },
+            bit_terms,
+            indices,
+        })
+    }
+}
+
+/// Check the pointer is not null and is aligned.
+fn check_ptr<T>(ptr: *const T) {
+    if ptr.is_null() {
+        panic!("Unexpected null pointer.");
+    };
+    if !ptr.is_aligned() {
+        panic!("Pointer is not properly aligned.");
+    };
+}
+
+/// Check a const pointer is not null and aligned, then cast it to a reference.
+///
+/// # Safety
+///
+/// This function is unsafe since it is reading from a raw pointer. Even if the pointer
+/// is not null and aligned, the memory could still be an invalid representation of the
+/// target type ``T``.
+unsafe fn const_ptr_as_ref<'a, T>(ptr: *const T) -> &'a T {
+    check_ptr(ptr);
+    let as_ref = unsafe { ptr.as_ref() };
+    as_ref.unwrap() // we know the pointer is not null, hence we can safely unwrap
+}
+
+/// Check a const pointer is not null and aligned, then cast it to a reference.
+///
+/// # Safety
+///
+/// This function is unsafe since it is reading from a raw pointer. Even if the pointer
+/// is not null and aligned, the memory could still be an invalid representation of the
+/// target type ``T``.
+unsafe fn mut_ptr_as_ref<'a, T>(ptr: *mut T) -> &'a mut T {
+    check_ptr(ptr);
+    let as_mut_ref = unsafe { ptr.as_mut() };
+    as_mut_ref.unwrap() // we know the pointer is not null, hence we can safely unwrap
+}
```

Comment:
Typically the "Safety" section tells you what you must do to call the function safely.  In this case, the comments could make it clearer that the null and aligned checks are handled safely, and the only `unsafe` bit of it is that the data _must_ be valid.  The only data that's truly unsafe in this sense is `BitTerm`, since all bit patterns of `u32`, `double complex` and `usize` are valid values of those types, they just might not be _correct_ values of those types.

Reply:
How about something like
```
/// # Safety
///
/// This function requires ``ptr`` to be non-null, aligned, initialized object of
/// type ``T``. While the resulting reference exists, the memory pointed to must not
/// be accessed otherwise.
```
(potentially strongly motivated of the `as_mut()` docstring) 😄

Reply:
in our case, it's safe to call the function with a null or misaligned pointer, because those will error safely. It's only the "must point to an initialised valid object of type `T`" bit that's unsafe in our version